### PR TITLE
fix(Core/DB): Fire Elemental Invasion changes

### PR DESCRIPTION
**Pull request by claudiodfc**
## Changes Proposed:
- Baron Charr added to Elemental Invasion game event (13)
- Blazing Invader's SAI rework.
- SMART_ACTION_SET_COUNTER extra parameter for subtracting values to the counter.
- Changed `Map::TempSummon* SummonCreature` to have WorldObjects as summoner and not just Units.
- Added a nullptr check for `SmartScript::GetTargets`

## Issues Addressed:

- Closes #8080 
https://github.com/chromiecraft/chromiecraft/issues/1440

## Tests Performed:

- Compiled and tested in Windows 10

## How to Test the Changes:

1. Go to ungoro and check the elemental Invasion
2. Only maximum of 3 Blazings should spawn at a time per Rift
3. After one is killed, after a while another spawns.

## Known Issues and TODO List:

None

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -1698,30 +1698,30 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
             }
         case SMART_ACTION_SET_COUNTER:
             {
-                if (ObjectList* targets = GetTargets(e, unit))
+            if (ObjectList* targets = GetTargets(e, unit))
```
Comment (Nyeriah) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:1701]: I might be wrong, but this probably should be four spaces to the right

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 2]
```diff
@@ -3707,25 +3707,41 @@ ObjectList* SmartScript::GetTargets(SmartScriptHolder const& e, Unit* invoker /*
...
                 {
-                    Unit* owner = l->front()->ToUnit();
-                    l->clear();
+                    if (Unit* owner = l->front()->ToUnit())
+                    {
+                        l->clear();
 
-                    if (Unit* base = ObjectAccessor::GetUnit(*owner, owner->GetCharmerOrOwnerGUID()))
-                        l->push_back(base);
+                        if (Unit* base = ObjectAccessor::GetUnit(*owner, owner->GetCharmerOrOwnerGUID()))
```
Comment (Nyeriah) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:3742]: missing brackets

File: src/server/game/AI/SmartScripts/SmartScript.h [hunk 3]
```diff
@@ -143,13 +143,17 @@ class SmartScript
...
     }
 
-    void StoreCounter(uint32 id, uint32 value, uint32 reset)
+    void StoreCounter(uint32 id, uint32 value, uint32 reset, uint32 subtract)
     {
         CounterMap::iterator itr = mCounterList.find(id);
         if (itr != mCounterList.end())
         {
-            if (reset == 0)
+            if (!reset && !subtract)
```
Comment (Nyeriah) [inline src/server/game/AI/SmartScripts/SmartScript.h:151]: missing brackets here too

**Comment by Nyeriah** [comment]
Changing summoner to worldobject is a natural approach, trinity did it too (IsSummonedBy()) and its something we probably should do for all functions, but it will take time

**Comment by IntelligentQuantum** [comment]
```
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/AI/SmartScripts/SmartScript.cpp: In member function ‘void SmartScript::ProcessAction(SmartScriptHolder&, Unit*, uint32, uint32, bool, const SpellInfo*, GameObject*)’:
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/AI/SmartScripts/SmartScript.cpp:1723:13: error: this ‘else’ clause does not guard... [-Werror=misleading-indentation]
 1723 |             else
      |  
```

**Comment by claudiodfc** [comment]
Yea there's an issue with indentation, I'll take care of it Monday.

**Review by UltraNix — APPROVED** [review]

**Comment by temperrr** [comment]
Looks good, respawn works,

however,

When I killed the bossof the invasion, server froze, eventually resulting in 
image

**Review by Nyeriah — APPROVED** [review]

**Comment by claudiodfc** [comment]
@temperrr can you re-test? I tried on my local and server didn't freeze for me. Maybe un-related to this PR?

**Comment by temperrr** [comment]
Cant get the event to start this time round.

event start 13 does nothing

**Comment by claudiodfc** [comment]
I'll see what I can do about that

**Comment by claudiodfc** [comment]
@temperrr Can you take a look? Added him to the `Elemental Invasions` event. None of the other elemental bosses are on the even so They'll have to be added on other PRs since this is for the Fire invasion.

**Comment by claudiodfc** [comment]
@azerothcore/testers please someone take a look

**Comment by temperrr** [comment]
image
image
[9c84b13a8bd3_worldserver.exe_[22-12_16-0-5].txt](https://github.com/azerothcore/azerothcore-wotlk/files/7763379/9c84b13a8bd3_worldserver.exe_.22-12_16-0-5.txt)

**Comment by claudiodfc** [comment]
It's strange how this doesn't crash mine, but it's obviously this PR. I'll have to take a look why tempsummon is crashing the server.

**Comment by claudiodfc** [comment]
@temperrr Made some changes, please take a look again when you got time.

It should no longer cause this freeze.

To further prove the change doesn't break the smart target, please execute this test bellow aswell:

#### Test case
.additem 49693 (Learn KT)
.cast 69731 (He laughs when hit by this spell)

**Comment by claudiodfc** [comment]
@Nyeriah @UltraNix I made a new change on the smart target, can you guys give it a look if It's ok?

**Comment by temperrr** [comment]
doesnt crash now :)

File: data/sql/updates/pending_db_world/rev_1637890270049187700.sql [hunk 4]
```diff
@@ -0,0 +1,21 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1637890270049187700');
+
+-- Adding Boss to Game event 13: Elemental Invasions
+DELETE FROM `game_event_creature` WHERE `eventEntry` = 13 AND `guid` = 14461;
+INSERT INTO `game_event_creature` (`eventEntry`, `guid`) VALUES ('13', '14461');
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1637890270049187700.sql:5]: quotes on ints. No.

**Review by Kitzunu — CHANGES_REQUESTED** [review]

**Review by Kitzunu — APPROVED** [review]

**Pull request merged**